src/reanimated2/layoutReanimation/sharedTransitions/SharedTransition.ts

```diff
@@ -0,0 +1,331 @@
+import { withTiming } from '../../animation';
+import type {
+  SharedTransitionAnimationsFunction,
+  SharedTransitionAnimationsValues,
+  CustomProgressAnimation,
+  ProgressAnimation,
+} from '../animationBuilder/commonTypes';
+import {
+  LayoutAnimationType,
+  SharedTransitionType,
+} from '../animationBuilder/commonTypes';
+import type { StyleProps } from '../../commonTypes';
+import {
+  configureLayoutAnimations,
+  registerEventHandler,
+  unregisterEventHandler,
+} from '../../core';
+import { Platform } from 'react-native';
+import { runOnUIImmediately } from '../../threads';
+import './ProgressTransitionManager';
+
+const supportedProps = [
+  'width',
+  'height',
+  'originX',
+  'originY',
+  'transform',
+  'borderRadius',
+];
+
+type AnimationFactory = (
+  values: SharedTransitionAnimationsValues
+) => StyleProps;
+
+type TransitionProgressEvent = {
+  closing: number;
+  goingForward: number;
+  eventName: string;
+  progress: number;
+  target: number;
+};
+
+const dummyWorklet = () => {
+  'worklet';
+  // noop
+};
+
+export class SharedTransition {
+  private _customAnimationFactory: AnimationFactory | null = null;
+  private _animation: SharedTransitionAnimationsFunction | null = null;
+  private _transitionDuration = 500;
+  private _customProgressAnimation?: ProgressAnimation = undefined;
+  private _progressAnimation?: ProgressAnimation = undefined;
+  private _defaultTransitionType?: SharedTransitionType = undefined;
+  private static _sharedElementCount = 0;
+  private static _eventHandler = {
```

Comment:
There seem to be too many loosely related things covered by this class. It both implements builder API and does things like registering for events. It feels like the event logic should be a part of "manager" class above. Maybe worth refactoring or even extracting this event logic or extracting the builder part as they all take a significant part of this file

Reply:
Resolved in: https://github.com/software-mansion/react-native-reanimated/pull/4421/commits/5ff4824408025f3e6ee6d862ee9ebbc18de98210